istioctl/cmd/proxystatus_v2.go

```diff
@@ -0,0 +1,151 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"text/tabwriter"
+	"time"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	corev2 "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/spf13/cobra"
+
+	"istio.io/istio/istioctl/pkg/clioptions"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/adsc"
+)
+
+const (
+	connectionsTypeURL = "istio.io/connections"
+)
+
+func newProxyStatusCommand() *cobra.Command {
+	var opts clioptions.CentralControlPlaneOptions
+
+	statusCmd := &cobra.Command{
+		Use:   "proxy-status [<pod-name[.namespace]>]",
+		Short: "Retrieves the synchronization status of each Envoy in the mesh",
+		Long: `
+Retrieves last sent and last acknowledged xDS sync from Pilot to each Envoy in the mesh
+
+`,
+		Example: `# Retrieve sync status for all Envoys in a mesh
+	istioctl proxy-status
+
+# Retrieve sync diff for a single Envoy and Pilot
+	istioctl proxy-status productpage-v1-679f4bcbbb-6j4j6.default
+`,
+		Aliases: []string{"ps"},
+		RunE: func(c *cobra.Command, args []string) error {
+			if opts.Xds == "" {
+				return fmt.Errorf("--endpoint default not implemented; MUST be supplied; try localhost:15012")
+			}
+			adscConn, err := adsc.Dial(opts.Xds, "", &adsc.Config{
+				IP: "1.2.3.4",
+				Meta: model.NodeMetadata{
+					Generator: "event",
+				}.ToStruct(),
+				// TODO I tried and failed to add DialOptions here and have ADSC
+				// pass them down so that I could add gRPC Interceptors for logging
+				// and debugging.
+			})
+			if err != nil {
+				return fmt.Errorf("Could not dial: %w", err)
+			}
+			if len(args) > 0 {
+				return fmt.Errorf("experimental proxy-status <pod> unimplemented")
+			}
+
+			err = adscConn.Send(&xdsapi.DiscoveryRequest{
+				Node: &corev2.Node{
+					Id: "sidecar~1.1.1.1~debug~cluster.local",
+				},
+				TypeUrl: connectionsTypeURL,
+			})
+			if err != nil {
+				return err
+			}
+
+			// TODO --timeout parameter
+			dr, err := adscConn.WaitVersion(10*time.Second, connectionsTypeURL, "")
+			if err != nil {
+				return err
+			}
+
```

Comment:
If you start connecting in 'regular' mode - adsc is parsing Listeners and Routes and Clusters, reproducing the requests that an envoy will make. It even splits the TCP vs HTTP listeners.

Reply:
@costinm I do not understand this comment.  Please expand.

Reply:
You get events by setting 'generator' to events. If you don't specify a generator - you'll get the LDS/RDS/CDS/EDS response - and adsc will parse it. 

Now that the new generator code seems to be working we can probably remove the extra generator meta - so one connection can handle any kind of XDS subscription.